### ISSUE TITLE
Non-Square adj_t for GCN

### ISSUE THREAD

**Issue by RX28666** [rusty1s/pyg_autoscale#29]
Hello,

I was trying to rewrite the code by myself. First, I found the outputted adj_t from SubgraphLoader is non-square, it has size`[target, target+neighbors]`. Hence, when I tried like:

`x = conv(x, adj)`

 it returns the AssertionError, because these lines in the gcn_convs:
```
if isinstance(edge_index, SparseTensor):
        assert edge_index.size(0) == edge_index.size(1)
```

I am wondering how the author solved this issue, and how to modify the code of SubgraphLoader in order to output square adj_t as normal case. Thanks!

**Comment by rusty1s** [comment]
You can do the GCN normalization outside of the layer. This should resolve the issue.

**Comment by RX28666** [comment]
Thanks for your reply! I implemented GAS in a naive way; the only difference is that I used the square adjacency matrix. However, I cannot reproduce the results using the same hyperparameters. I am wondering if the shape of the adjacency matrix will affect the performance or if there are any other errors in my implementation. Here is my implementation:

```
class Net(nn.Module):
    def __init__(self, num_features, hidden_channels, num_classes, num_layers, num_nodes, **kwargs):
        super(Net, self).__init__()
        
        self.convs = torch.nn.ModuleList()
        self.convs.append(GCNConv(in_channels, hidden_channels))
        self.bns = torch.nn.ModuleList()
        self.bns.append(torch.nn.BatchNorm1d(hidden_channels))
        for _ in range(num_layers - 2):
            self.convs.append(GCNConv(hidden_channels, hidden_channels))
            self.bns.append(torch.nn.BatchNorm1d(hidden_channels))
        self.convs.append(GCNConv(hidden_channels, out_channels))
        self.history = []
        self.history.append(torch.zeros(num_nodes, hidden_channels).to(device))
        for _ in range(num_layers - 2):
            self.history.append(torch.zeros(num_nodes, hidden_channels).to(device))

    def reset_parameters(self):
        for conv in self.convs:
            conv.reset_parameters()
        for bn in self.bns:
            bn.reset_parameters()

    def forward(self, x, adj, id, size):
        for i, conv in enumerate(self.convs[:-1]):
            x = conv(x, adj)
            x = self.bns[i](x)
            x = F.relu(x)
            self.history[i][id[:size]] = x[:size].clone().detach()
            x_cat = torch.cat([x[:size], self.history[i][id[size:]]])
            x_cat = F.dropout(x_cat, p=args.dropout, training=self.training)    
        x_out = self.convs[-1](x_cat, adj)
        out = F.log_softmax(x_out, dim=1)
        return out
```

```
for i, batch in enumerate(train_loader):
           
        x = batch.x.to(device)
        adj_t = batch.adj_t.to(device)
        id = batch.n_id.to(device)
        optimizer.zero_grad()
        train_mask = batch.train_mask[:batch.batch_size]
        y = batch.y[:batch.batch_size][train_mask].to(device)
        out = model(x, adj_t, id, batch.batch_size)[:batch1.batch_size][train_mask]
        loss = F.nll_loss(out, y)
        loss.backward()
        optimizer.step()
```
Any help would be appreciated.

**Comment by rusty1s** [comment]
You need to initialize `GCNConv` with `normalize=False`.
